### ISSUE TITLE
how can one server handle more than one client？

### ISSUE THREAD

**Issue by mbigger** [mz-automation/lib60870#1]
I have a problem，if i have more than one client，and these clients have to bind on the same sever which is in a  specific host.
How to write the server-end program,how to send the datas,what's the best way to use this library?

**Comment by mzillgith** [comment]
The protocol is not intended to have more than one active client (master). There can be multiple TCP clients connected in standby but only one client is "active" an can send/receive application layer data. This protocol can be seen as a replacement for the balanced master/slave connection as specified in 60870-5-101 that has always only one master and only one slave. If you want to serve multiple clients you need multiple servers running on different ports.

**Comment by pavel-ch** [comment]
Hi Michael,
this looks like a misconception to mee. It is quite often and common to communicate in the "one server -
 many clients" model in 104 networks with the server on one default port 2404. Imagine a substation concentrator device, for example, which has to communicate to the dispatch centre and to the local visualization device, both on 104. Only if there are some connections required (and defined)  in a redundant mode, there are some standby connections.
Br Pavel

**Comment by mzillgith** [comment]
Hi Pavel,
thanks for your hint. In general the specification states that there is only one started connection where ASDUs are sent (also in the test case descriptions in 604 part). However there is the concept of "redundancy groups" (a set of connections with dedicated process image and event buffer) in the standard. But this is only vaguely defined and not yet supported by the library. Is this concept used in the use cases you were talking about?

**Comment by pavel-ch** [comment]
Yes I mentioned these redundancy groups in the last sentence above. I have not met this concept implemented in practice.
But generally, not considering these redundant connections, the 104 server (slave) should allow connections from multiple clients. Nearly all 104 devices known to me allow this mode. It is quite common that 104 server has a whitelist configuration for filtering clients (client IP addresses) allowed to connect to. All client are handled equally, usually all ASDUs (e.g. COT=3 and similar) generated by the server are sent to all actually connected clients (with an exception of COT=20 or COT=7/10 ASDUs etc. initiated by a particular client - these are sent to that client only).

**Comment by mzillgith** [comment]
Ok. Seems reasonable. This way each client connection is conceptionally its own redundancy group with an own event buffer.

**Comment by mbigger** [comment]
Additionally，the multiple clients may have different configures，for example，one's common address is 5, another is 10 . Also the ASDUs generated  by the server for one client may be different from these for another.
But all the clients distributed  in different place connect to the same server IP+Port.
It is a real  problem we have solved in our project，because often we need to send data to different 
dispatch centre，but the datas they required are quite different.

**Comment by mzillgith** [comment]
I think I can easily add a new server mode where each connected client has its own "redundancy group", by sending/receiving I/S messages and receiving the same events from the server. When you also need different data maps and different events then I think a more complex configuration is required and also the server side API has to be changed. I will think about that. Any idea is welcome.

**Comment by mzillgith** [comment]
The server/slave supports now two modes. The existing mode (SINGLE_REDUNDANCY_GROUP) that allow only one "active" connection and a single event buffer that also stores events when no client is connected, and a new mode (CONNECTION_IS_REDUNDANCY_GROUP) that allow multiple clients to be "active". For the latter a new dedicated event buffer is created when the client connects.

**Comment by artyompetrov** [comment]
HI! Is that correct that when server.ServerMode = ServerMode.SINGLE_REDUNDANCY_GROUP;
its keeps sending START DT and answering to interogation commands on the second connection?

As i understood, according to your code it is possible to have two connections with property IsActive = true; (because the connection becomes active when it sends START DT)

**Comment by artyompetrov** [comment]
And you had an error in Bitstring32.cs, it wasn't possible to send a Birstring32 because QualityDescriptor equals null (Don't know if it fixed already)

**Comment by mzillgith** [comment]
BitString32 should be fixed https://github.com/mz-automation/lib60870/commit/256bcb5a467ce940cc26c1667c24e1a72d199f71

**Comment by mzillgith** [comment]
> As i understood, according to your code it is possible to have two connections with property IsActive = true; (because the connection becomes active when it sends START DT)

Yes I see. This is a bug. The other connections should be set to be inactive when a connection receives START_DT

**Comment by artyompetrov** [comment]
> Yes I see. This is a bug. The other connections should be set to be inactive when a connection receives START_DT

Maybe it's better to decline all new connections (don’t send START DT con) if there is already one active connection. Otherwise if two clients will connect to one server, the server will constantly jump from one connection to another. At least the meter with IEC 104 that I have acts like this. So it's possible to have few simultaneous TCP connections and as soon as the active connection becomes inactive another connection waiting in the line becomes active.

**Comment by mzillgith** [comment]
I think the behaviour you described is not intended by the standard, because the controlling station (client) should control which connection is used. Also the standard days: "If the STARTDT is not confirmed, the connection is closed by the controlling station." But the correct behaviour of the controlling station would be to send STOPDT on one connection and only after doing that sending a STARTDT on the other connection.

**Comment by artyompetrov** [comment]
I was wrong about keeping TCP connection active, you are right, if a client doesn't receive the START DT it closes connection.
The reason why I think something is wrong with your algorithm is that it is possible to destroy and intercept the flow of telemetry, just by connecting a new client. It doesn't work like this in any IEC 104 implementation that I am familiar with.

The meter that only accepts one connection at a time it works like this:
If the meter's IEC 104 server already has an active connection it just doesn't answer new connection’s START DT until the active connection is closed so all the new clients fail to connect.
But if an inactive client (or more) hasn't closed the TCP connection (T1 time isn't over yet) the server sends START DT CON just after the active connection becomes inactive (in case of several inactive connections, the first connection in the queue becomes active).

A SCADA system (most popular in Russia) which accepts multiple connections at a time works like this:
If the server already has an active connection it accepts new connections (sends START DT CON), so the clients don't automatically close the TCP connection, but doesn't sends telemetry and doesn't answer interrogation commands (but answer test commands). As soon as the active connection is closed, an inactive connection waiting in line becomes active.

So I think it is better to implement the algorithm in one of these ways (or both) in addition to ServerMode.CONNECION_IS_REDUNDANCY_GROUP;

**Comment by mzillgith** [comment]
Thank you for the information about the different treatment of the problem in implementation you mentioned. This is a very interesting discussion. I guess the reason why implementations behave so different is that the specification was not so clear about that in the past. The latest specifications contains more details about the concept of the redundancy groups.

**Comment by burrj** [comment]
I agree that this is a very interesting conversation and a topic that has confused me for some time.  I have no experience with 104 until now and wondered how existing servers handle multiple (non-redundant) clients and also how often this is done.  
Would a typical PLC device that is a 104 server typically support multiple non-redundant clients?
Would a typical PLC device that is a 104 server typically support redundant connections?

It seems that the 104 specification assumed a single master until the redundancy section was added.  After reading the redundancy section, you can apply some of what it says to non-redundant connections also.  Specifically, the redundancy section says that a redundancy group must use only one process image (database/event buffer). It also says that if more that one client needs to access the server at the same time, the clients must be assigned to different process images.  I think "process image" means the same thing as "sector" which means the same thing as "database" but cannot find an official definition of any of these terms.

I would rather not support multiple sectors just to support multiple (non-redundant) clients, but that seems to be the expectation.  A server with two sectors is almost like two different servers (or two Logical RTUs as the 101 specification calls them in sections 7.4.1 and 7.4.5).  If having two sectors is almost like being 2 servers, it easy then easy to handle two different clients.  This highlights that the original intention seemed to be that a single server supported a single client.

But I have no 104 experience.  Does this thinking align with what you see in the real world?

Also, Figure 1 of the 104 specification shows only a single client and single server, with redundancy needed because of the different routers and the physical communication paths between them.  A client needs to initiate the TCP/IP connection to the server, and TCP/IP provides no direct way to specify the routing used by the connection.  Does a server that supports redundancy listen on different TCP/IP addresses so that the TCP/IP routing layer allows the client to establish connections along those different physical paths?  There would be little use in having redundant connections if all the connections used the same physical media.  Or do I misunderstand the diagram and intent of redundancy?

**Comment by mzillgith** [comment]
I am thinking in implementing a redundancy group as a number of client connections that share the same event buffer. I think the main function is not to loose events when there is a problem in the network. At the moment I don't think about having completely different "process images" for the different redundancy groups as I think the handling from application point of view would be to complicated. Also at the moment I don't see a use case. Please let me know if you see a useful application for that?

**Comment by burrj** [comment]
I personally have no experience with 101 or 104 but I think you are correct: the protocol was only designed for one master per slave.  It seems that many slaves (servers) support multiple masters (clients).  I have recently read many product manuals but few of them say how many clients the server supports.  Many IEC 104 servers DO support having different "process images" that are usually called "sectors" in the product documentation.  Sections 7.4.1 and 7.4.5 of the IEC 60870-5-101 specification use the term "sector" and also "logical RTU" or "LRU" and provide a hint of the use case that is pictured in figure 90 of the 101 specification.  If a server is acting as a data concentrator for multiple RTUs it can represent each RTU as a sector. 

I have read a document by Statkraft that states that on the server, the process images must be different and individually configured per redundancy group and may contain different sets of objects.  The document does not really provide a use case.  I read somewhere that different clients could be interested in different information objects and having different process images (each with a unique COA) allows each client to receive only the events of interest.  This makes sense to me but is putting all of the burden on the server.  The clients could also simply ignore events for objects that they do not use.

The other reason to support multiple "process images" is that the redundancy section of the 104 specification says "If more than one controlling station need to access the same controlled station simultaneously, each controlling station must be assigned to a different redundancy group
(process image)."
